lib/Sema/CompilerRepresentable.cpp

```diff
@@ -0,0 +1,325 @@
+    break;
+  case ResultUnrepresentable:
+    assert(UnderlyingReason);
+    TC.diagnose(Target, diag::compiler_evaluable_unrepresentable_result,
+                UnrepresentableType, UnderlyingReason->UnrepresentableType);
+    break;
+  }
+
+  assert(UnderlyingReason);
+  UnderlyingReason->emitDiagnosticNotesImpl(TC, Target);
+}
+
+void UnrepresentableReason::emitDiagnosticNotes(TypeChecker &TC,
+                                                const Decl *Target) const {
+  emitDiagnosticNotesImpl(TC, Target);
+}
+
+void UnrepresentableReason::emitDiagnosticNotes(TypeChecker &TC,
+                                                SourceLoc Target) const {
+  emitDiagnosticNotesImpl(TC, Target);
+}
+
+// Checks if the compile-time evaluator can handle the types that the generic
+// arguments are bound to.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkGenericArgs(CanBoundGenericType type) {
+  assert(type);
+  for (auto arg : type->getGenericArgs()) {
+    if (auto unrepresentable = check(arg)) {
+      return UnrepresentableReason::getGenericArgUnrepresentable(
+          type, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle this enum.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkDeclaredType(const EnumDecl *decl) {
+  assert(decl);
+  for (auto *elementDecl : decl->getAllElements()) {
+    auto *parameterList = elementDecl->getParameterList();
+    if (!parameterList)
+      continue;
+
+    for (auto *paramDecl : *parameterList) {
+      if (auto unrepresentable = check(paramDecl->getType())) {
+        return UnrepresentableReason::getEnumElementUnrepresentable(
+            decl->getDeclaredType(), elementDecl, std::move(unrepresentable));
+      }
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle this struct.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkDeclaredType(const StructDecl *decl) {
+  assert(decl);
+  for (auto *fieldDecl : decl->getStoredProperties()) {
+    if (auto unrepresentable = check(fieldDecl->getType())) {
+      return UnrepresentableReason::getFieldUnrepresentable(
+          decl->getDeclaredType(), fieldDecl, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle the types of the parameters
+// of this function.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkParams(CanAnyFunctionType type) {
+  assert(type);
+  for (auto param : type->getParams()) {
+    if (auto unrepresentable = check(param.getType())) {
+      return UnrepresentableReason::getParamUnrepresentable(
+          type, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle values of type `type`.
+std::unique_ptr<const UnrepresentableReason>
```

Comment:
High-level comment: the use of `std::unique_ptr` feels a bit heavyweight for the purpose here. Would it be better modeled like [`swift::SymbolicValue`](https://github.com/apple/swift/blob/7235731d001f8fe264efd9dd79a6940ce8b9edfd/include/swift/SIL/SILConstants.h#L66)?

Reply:
I need some kind of pointer somewhere because `UnrepresentableReason` refers to itself in the `UnderlyingReason` field. `SymbolicValue` does not do this (yet? it might be necessary for producing "stack traces").

I thought of some of ways to deal with that pointer, and I'm not sure which one is best because I'm not very experienced with C++ yet. Some ways:
* What I did with `unique_ptr`.
* Allocate inside the ASTContext's allocator, so that the ASTContext owns it, and use a bare pointer.
* Always return `UnrepresentableReason`s by value, so that clients do not have to deal with pointers. Deal with the allocations and pointers inside of `UnrepresentableReason`'s constructors.

Reply:
Did you consider using `BumpPtrAllocator` in your checker pass, and use that to allocate during checking instead of using the AST context? After checking and diagnosing, the allocator can be simply thrown away.

Reply:
UnrepresentableReason only gets actually allocated in the case of a compile-time error, right?  If so, I think it is perfectly reasonable to use this sort of representation.  We don't generally care about low-level performance like this in the error cases, we prefer to burn cycles to generate good error messages - I like your attention to detail here!

I'd suggest a typedef for the 'std::unique_ptr<const UnrepresentableReason>' type tho!

SymbolicValue is different in this case because tons of them are allocated and passed around in the normal case.  that is why its inline representation optimization is important.

Reply:
yep, there's only an allocation when there is an error.

I have made a typedef.